src/libs/types-rs/src/container.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) 2019-2021 Alibaba Cloud
+// Copyright (c) 2019-2021 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::fmt::{Display, Formatter};
+use std::str::FromStr;
+
+const POD_CONTAINER: &str = "pod_container";
+const POD_SANDBOX: &str = "pod_sandbox";
+
+const STATE_READY: &str = "ready";
+const STATE_RUNNING: &str = "running";
+const STATE_STOPPED: &str = "stopped";
+const STATE_PAUSED: &str = "paused";
+
+/// Error codes for container related operations.
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    /// Invalid container type
+    #[error("Invalid container type {0}")]
+    InvalidContainerType(String),
+    /// Invalid container state
+    #[error("Invalid sandbox state {0}")]
+    InvalidState(String),
+    /// Invalid container state transition
+    #[error("Can not transit from {0} to {1}")]
+    InvalidStateTransition(State, State),
+}
+
+/// Types of pod containers: container or sandbox.
+#[derive(PartialEq, Debug, Clone)]
+pub enum ContainerType {
+    /// A pod container.
+    PodContainer,
+    /// A pod sandbox.
+    PodSandbox,
+}
+
+impl ContainerType {
+    /// Check whether it's a pod container.
+    pub fn is_pod_container(&self) -> bool {
+        matches!(self, ContainerType::PodContainer)
+    }
+
+    /// Check whether it's a pod container.
+    pub fn is_pod_sandbox(&self) -> bool {
+        matches!(self, ContainerType::PodSandbox)
+    }
+}
+
+impl Display for ContainerType {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        match self {
+            ContainerType::PodContainer => write!(f, "{}", POD_CONTAINER),
+            ContainerType::PodSandbox => write!(f, "{}", POD_SANDBOX),
+        }
+    }
+}
+
+impl FromStr for ContainerType {
+    type Err = Error;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        match value {
+            POD_CONTAINER => Ok(ContainerType::PodContainer),
+            POD_SANDBOX => Ok(ContainerType::PodSandbox),
+            _ => Err(Error::InvalidContainerType(value.to_owned())),
+        }
+    }
+}
+
+/// Process states.
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum State {
+    /// The container is ready to run.
+    Ready,
+    /// The container executed the user-specified program but has not exited
+    Running,
+    /// The container has exited
+    Stopped,
+    /// The container has been paused.
+    Paused,
+}
+
+impl Display for State {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        match self {
+            State::Ready => write!(f, "{}", STATE_READY),
+            State::Running => write!(f, "{}", STATE_RUNNING),
+            State::Stopped => write!(f, "{}", STATE_STOPPED),
+            State::Paused => write!(f, "{}", STATE_PAUSED),
+        }
+    }
+}
```

Comment:
You could use https://crates.io/crates/strum rather than creating the `const` values and this implementation.

Reply:
How about refining this after merging?